main.cpp

```diff
@@ -370,196 +304,44 @@ int main()
     std::cout << "intNum: " << intNum << std::endl;
     
     {
-        using Type = decltype(f)::Type;
-        f.apply([&f](std::unique_ptr<Type>&value) -> decltype(f)&
-                {
-                    auto& v = *value;
-                    v = v * v;
-                    return f;
-                });
+        using Type = decltype(f);
+        f.apply([&f](std::unique_ptr<Type::Primitive>& a) -> Type&
+            {
+                auto& v = *a;
+                v = v * v;
+                return f;
+            });
```

Comment:
You are not supposed to change main()
You are supposed to change your class to make main() work.
- [ ] put it back the way it was, change your template class so it compiles again.